Add hybrid spa flow support (#3978)
AzureAD/microsoft-authentication-library-for-js: samples/msal-browser-samples/HybridSample/README.md

> # Hybrid SPA sample application

Reviewer: We should also link to this sample from the browser and node READMEs and FAQs

Author: Good call.

This sample demonstrates how to use MSAL.js v2 and MSAL Node together in a "hybrid" application that performs both server-side and client-side authenication. 

> It shows how to use two new APIs, `returnSpaCode` in MSAL Node and `acquireTokenByCode` in MSAL.js v2, to authenicate a user server-side using a confidential client, and then SSO that user client-side using a second authorization code that is returned to the confidential client and redeemed by the public client client-side. This helps mitigate user experience and performance concerns that arise when performing server-side and client-side authenication for the same user, especially when third-party cookies are blocked by the browser.

Author: ```suggestion
It shows how to use two new APIs, `enableSpaAuthCode` in MSAL Node and `acquireTokenByCode` in MSAL.js v2, to authenticate a user server-side using a confidential client, and then SSO that user client-side using a second authorization code that is returned to the confidential client and redeemed by the public client client-side. This helps mitigate user experience and performance concerns that arise when performing server-side and client-side authentication for the same user, especially when third-party cookies are blocked by the browser.
```

## Setup

1. In the Azure Portal, create a new app registration.
2. In the root folder for this sample, create a `.env` file and add the client id for this application as `MSAL_CLIENT_ID`.
3. In the Azure Portal, under the **Authentication** tab for your application, add the following **Web** redirect URIs:
    1. `http://localhost:3000/auth/server-redirect`
    2. `http://localhost:3000/auth/implicit-redirect`
4. Also add the following **Single-page application** redirect URIs:
    1. `http://localhost:3000/auth/client-redirect`
5. Under **Implicit grant and hybrid flows**, check the boxes to enable **Access tokens** and **ID tokens**.
6. Under the **Certificats & secrets** tab, create a new client secret. Add this client secret to the `.env` file as `MSAL_CLIENT_SECRET`.

> 7. In the manifest editor, add the following optional ID token claims:

Reviewer: Might be good to include images for some of these portal steps

Author: Pictures are annoying to maintain, will stay with text for now.

    auth: {
        clientId: process.env.MSAL_CLIENT_ID,
        authority: "https://login.microsoftonline.com/common",
        clientSecret: process.env.MSAL_CLIENT_SECRET
    },
    system: {
        loggerOptions: {
            loggerCallback: (loglevel, message, containsPii) => {
                console.log(message);
            },
            piiLoggingEnabled: false,
            logLevel: msal.LogLevel.Verbose,
        }
    }
});
```

Next, generate an auth code url and navigate the user:

Source: [routes/auth.js](./routes/auth.js)
```typescript
router.get('/login', (req, res) => {
    const authCodeUrlParameters = {
        scopes: ["user.read"],
        redirectUri: "http://localhost:3000/auth/server-redirect",
        responseMode: "form_post"
    };

    // Generate auth code url and redirect the user
    msalInstance.getAuthCodeUrl(authCodeUrlParameters)
        .then((response) => {
            console.log(response);
            res.redirect(response);
        })
        .catch((error) => console.log(JSON.stringify(error)));
});
```

Next, parse the authorization code, and invoke the `acquireTokenByCode` API on the `ConfidentialClientApplication` instance. 

> When invoking this API, set `returnSpaCode` to `true`, which will enable MSAL to acquire a second authorization code to be redeemed by your single-page application. 

Author: ```suggestion
When invoking this API, set `enableSpaAuthCode` to `true`, which will enable MSAL to acquire a second authorization code to be redeemed by your single-page application. 
```

Your application should parse this second authorization code, as well as any account hints (e.g. `sid`, `login_hint`, `preferred_username`) and return them such that they can be rendered client-side:

Source: [routes/auth.js](./routes/auth.js)
```typescript
router.post('/server-redirect', (req, res) => {
    const tokenRequest = {
        code: req.body.code,
        scopes: ["user.read"],
        redirectUri: "http://localhost:3000/auth/server-redirect",

>         returnSpaCode: true

Author: ```suggestion
        enableSpaAuthCode: true
```

        })
        .catch((error) => {
            console.timeEnd(timeLabel)
            console.log(error);
            res.status(500).send(error);
        });
});
```

### Public client

First, configure a new `PublicClientApplication` from MSAL.js in your single-page application:

Source: [views/client-redirect.hbs](./views/client-redirect.hbs)
```typescript
const msalInstance = new msal.PublicClientApplication({
    auth: {
        clientId: "{{clientId}}",
        redirectUri: "http://localhost:3000/auth/client-redirect",
        authority: "https://login.microsoftonline.com/common"
    }
})
```

Next, render the `code` that was acquired server-side, and provide it to the `acquireTokenByCode` API on the MSAL.js `PublicClientApplication` instance. Be sure to not include any additional scopes that were not included in the first login request, otherwise the user may be prompted for consent.

The application should also render any account hints, as they will be needed for any interactive requests to ensure the same user is used for both requests.

Source: [views/client-redirect.hbs](./views/client-redirect.hbs)
```typescript
const sid = "{{sid}}";
const code = "{{code}}";

> const preferredUsername = "{{preferredUsername}}";

Reviewer: `sid` and `preferredUsername` aren't used below. Might create confusion.

Author: Good catch.

# Hybrid SPA sample application

> This sample demonstrates how to use MSAL.js v2 and MSAL Node together in a "hybrid" application that performs both server-side and client-side authenication. 

Reviewer 2: Can we change this description to also reflect the ssoSilent working with an auth code without depending on a "hint"?

Author: Not sure what you mean exactly....are you saying to call out that this can be used instead of `ssoSilent` with login hint?
